internal/v5/auth.go

```diff
@@ -110,43 +110,13 @@ func (h *ReqHandler) authorizeEntityAndTerms(req *http.Request, entityIds []*rou
 		return authorization{}, errgo.WithCausef(nil, params.ErrUnauthorized, "entity id not specified")
 	}
 
-	ACLs := make([][]string, len(entityIds))
-	requiredTerms := make(map[string]struct{})
-	allowAuth := true
-	for i, entityId := range entityIds {
-		entity, err := h.Store.FindEntity(entityId)
-		if err != nil {
-			return authorization{}, errgo.Mask(err, errgo.Is(params.ErrNotFound))
-		}
-		baseEntity, err := h.Store.FindBaseEntity(&entityId.URL, "acls", "developmentacls")
-		if err != nil {
-			return authorization{}, errgo.Mask(err, errgo.Is(params.ErrNotFound))
-		}
-
-		ACLs[i] = baseEntity.ACLs.Read
-		if entityId.Development {
-			ACLs[i] = baseEntity.DevelopmentACLs.Read
-		}
-
-		if entity.CharmMeta == nil || len(entity.CharmMeta.Terms) == 0 {
-			// No need to authenticate if the ACL is open to everyone.
-			open := false
-			for _, name := range ACLs[i] {
-				if name == params.Everyone {
-					open = true
-					break
-				}
-			}
-			allowAuth = allowAuth && open
-		} else {
-			allowAuth = false
-			for _, term := range entity.CharmMeta.Terms {
-				requiredTerms[term] = struct{}{}
-			}
-		}
+	public, acls, requiredTerms, err := h.entityAuthInfo(entityIds)
+	if err != nil {
+		return authorization{}, errgo.Mask(err)
 	}
+
 	// if all entities are open to everyone and non of the entities defines any Terms, then we return nil
-	if allowAuth {
+	if public {
```

Comment:
A publicly available charm with terms should still require agreement to them. Shouldn't this be `if public && len(requiredTerms) == 0`?